File: 200020120_200050129_200050157/code/simulator.cpp
#include "include/header.hpp"
#include <random>


/*!
 * @brief Constructor for the simulator class It acts as a Global scope for all the other functions
 *
 * @param seed Random seed
 * @param z0 Fractioin of Slow users
 * @param z1 Fraction of lowCpu Users
 * @param Ttx Inter Transaction Time
 * @param min_ngbrs Min number of neighbors
 * @param max_ngbrs Max number of Neighbors
 * @param sim_time The Simulation time Limit
 */
simulator::simulator(int seed, ld z0, ld z1, ld Ttx, ld Tblk, int min_ngbrs, int max_ngbrs, ld sim_time) {

	// Set the seeds
    this->seed = seed;
    rng.seed(seed);
    rng_64.seed(seed);


	// Store the Parameters
    this->z0 = min(1.0L, max(0.0L, z0));    // fraction in [0,1]
    this->z1 = min(1.0L, max(0.0L, z1));
    this->Ttx = Ttx;
    this->Tblk = Tblk;
    this->Simulation_Time = sim_time;

    // bits per second
    fast_link_speed = 100*(1<<20);          // 100 Mbps
    slow_link_speed = 5*(1<<20);            // 5 Mbps
    queuing_delay_numerator = 96*(1<<10);   // 96 kbps
	

	// Use the pick_random function to generate indices with slow/lowCPU peers
    vector<int> slow_indices = pick_random(n, z0*n);
    vector<int> lowCPU_indices = pick_random(n, z1*n);

	// Create a vector containing all the peers(nodes)
    peers_vec.reserve(n);

	for(int i = 0; i < n; i++){
		peers_vec.push_back(peer(i));
	}

    for(int& i:slow_indices) {
        peers_vec[i].slow = true;
    }
    for(int& i:lowCPU_indices) {
        peers_vec[i].lowCPU = true;
    }
    for (int i=0; i<n; i++) { 		// calculate fraction of the hashing power depending on the Speed of peer
        if(peers_vec[i].lowCPU) peers_vec[i].fraction_hashing_power = 1.0L/(10.0L*(n-lowCPU_indices.size())+lowCPU_indices.size());
        else peers_vec[i].fraction_hashing_power = 10.0L/(10.0L*(n-lowCPU_indices.size())+lowCPU_indices.size());
    }

    for (int i=0; i<n; i++) {

		// Set up the first transaction generation event for each peer
        ld time_txn = exponential_distribution<ld>(1.0L/Ttx)(rng);
        event* e = new event(time_txn, 1, &peers_vec[i]);
        this->push(e);

		// Set up the first block generation event for each peer
		ld time_blk = exponential_distribution<ld>(peers_vec[i].fraction_hashing_power/this->Tblk)(rng);
        event* blk_gen = new event(time_blk, 4, &peers_vec[i]);
        this->push(blk_gen);
    }


	// Initialize the adjacency matrix
    adj = vector<vector<int>>(n);
    rho = vector<vector<ld>>(n, vector<ld>(n, 0.0L));
    visited = vector<bool>(n, false);


	// Call the create_graph method to create a random graph
    this->create_graph(min_ngbrs, max_ngbrs);

    blk::blk_id_to_blk_ptr.insert(make_pair(peer::genesis->blk_id,peer::genesis));
}



/**
 * @brief Utility Function to choose k random numbers from 0 to n-1
 */
vector<int> simulator::pick_random(int n, int k) {
    // https://stackoverflow.com/a/28287865/17786040

    k = min(n, k);

    unordered_set<int> elems;

    for (int r=n-k; r<n; r++) {
        int v = uniform_int_distribution<>(0, r)(rng);  // [0,r]
        if (!elems.insert(v).second) {
            elems.insert(r);
        }
    }

    // deterministic order, so shuffle it
    vector<int> result(elems.begin(), elems.end());
    shuffle(result.begin(), result.end(), rng);

    return result;
}

/**
 * @brief Function to traverse a graph in dfs manner 
*/
void simulator::dfs(int node, vector<unordered_set<int>>& adj_sets) {
    visited[node] = true;
    for(int i:adj_sets[node]) {
        if(!visited[i]) {
            dfs(i, adj_sets);
        }
    }
}
/**
 *	@brief create a random connected graph,
 *  where each node has min_ngbrs <= #neighbors <= max_ngbrs
 *
*/
void simulator::create_graph(int min_ngbrs, int max_ngbrs) {

    vector<unordered_set<int>> adj_sets(n);

    while(true) {
        fill(visited.begin(), visited.end(), false);
        dfs(0, adj_sets);
        bool connected = true;
        for (int i=0; i<n; i++) {
            if(!visited[i]) {
                connected = false;
                break;
            }
        }
        if(connected) break;

		// Try a different graph
        while(true) {
            bool done = true;
            for (int i=0; i<n; i++) {
                adj_sets[i].clear();
            }
            for (int i=0; i<n; i++) {

				// choose the number of neighbours for each node
                int req_ngbrs = uniform_int_distribution<>(min_ngbrs,max_ngbrs)(rng) - adj_sets[i].size();
                if(req_ngbrs<0) {
                    done = false;
                    break;
                }

				// use pickrandom to choose req_ngbrs peers
                vector<int> new_ngbrs = pick_random(n, req_ngbrs);

				// for each chosen neigbhour check the constraints and add the neighbour to the adjacency matrix
                for (int j=0; j<new_ngbrs.size(); j++) {
                    if(new_ngbrs[j]==i || adj_sets[new_ngbrs[j]].size()>=max_ngbrs) {
                        done = false;
                        break;
                    }
                    adj_sets[i].insert(new_ngbrs[j]);
                    adj_sets[new_ngbrs[j]].insert(i);     // undirected graph
                }

				// check the min_ngbrs constraint
                for (int j=0; j<n; j++) {
                    if(adj_sets.size()<min_ngbrs) {
                        done = false;
                        break;
                    }
                }
                if(done==false) break;
            }
            if(done) break;
        }

    }

    for (int i=0; i<n; i++) {
        adj[i] = vector<int>(adj_sets[i].begin(), adj_sets[i].end());
    }


	// set up a latency between all neighbourse
    for (int i=0; i<n; i++) {
        for(auto&& j:adj[i]) {
            rho[i][j] = (10.0L + (490.0L * uniform_real_distribution<ld>()(rng))) * 0.001;
        }
    }

}

/**
 * @brief Utility function to print the adjacency matrix
 */
void simulator::print_graph() {
    // print graph/network adjacency list
    for (int i=0; i<n; i++) {
        cout << i << ": ";
        for(auto&& j:adj[i]) {
            cout << j << ", ";
        }
        cout << '\n';
    }
}

/**
 * @brief Take the first event in the priority queue, run it till the queue is empty or timlimit is exceeded
 */
void simulator::run() {
    // https://www.cs.cmu.edu/~music/cmsip/readings/intro-discrete-event-sim.html

    while(!pq_events.empty()) {
        // cout << pq_events.size() << '\n';
        event* e = pq_events.top();
        pq_events.pop();
        
        if(e->timestamp > Simulation_Time && (e->type==1 || e->type==4)) continue;

        e->run(*this);
        
        // clean up (delete event, and probably the associated txn)
        // (define destructor for event)
    }
}

/**
 * @brief Add event e to the priority queue
 */
void simulator::push(event* e) {
    pq_events.push(e);
}

/**
 * @brief Output the whole block chain after it has settled down
 *
 * @param out Output stream to which it is printed
 */
void simulator::print_entire_tree(ostream& out) {
    for (ll i=0; i<blk::curr_blk_id; i++) {
        auto it = blk::blk_id_to_blk_ptr.find(i);
        if(it == blk::blk_id_to_blk_ptr.end() || it->second==nullptr) continue;

        for(blk* child:it->second->children) {
            if(child==nullptr) continue;
            out << "id" << it->first << "miner";
            if(it->second->miner == nullptr) out << "G";
            else out << it->second->miner->id;
            out << "invalid" << it->second->originally_invalid;
            out << " ";
            out << "id" << child->blk_id << "miner" << child->miner->id << "invalid" << child->originally_invalid << endl;
        }
    }
}

File: python/simulator.py
'''Import defaults, Blk, Event and Peer'''
import random
from io import TextIOWrapper
import numpy as np
from block import Blk
from peer import Peer
from event import Event

class Simulator:
    '''Simulator'''

    def __init__(self, n, seed, z0, z1, Ttx, Tblk, m, M, T):
        self.n = n
        self.seed = seed
        random.seed(seed)
        np.random.seed(seed)
        self.z0 = min(1.0,max(0.0,z0))
        self.z1 = min(1.0,max(0.0,z1))
        self.Ttx = Ttx
        self.Tblk = Tblk
        self.T = T
        self.m = m
        self.M = M
        self.fls = 100 * (2 ** 20)
        self.sls = 5 * (2 ** 10)
        self.qdn = 96 * (2 ** 10)
        self.slow = np.random.choice(n,int(z0 * n))
        self.low = np.random.choice(n,int(z1 * n))
        num_low = len(self.low)
        norm_fac = 10.0 * (n - num_low) + num_low
        self.peers = []
        self.prioq = []
        for i in range(n):
            self.peers.append(Peer(n,i))
            if i in self.slow:
                self.peers[i].slow = True
            if i in self.low:
                self.peers[i].low = True
                self.peers[i].alpha = 10.0 / norm_fac
            else:
                self.peers[i].alpha = 1.0 / norm_fac
        for i in range(n):
            time_txn = np.random.exponential(self.Ttx)
            txn_eve = Event(time_txn,1,self.peers[i])
            self.push(txn_eve)
            time_blk = np.random.exponential(self.Tblk)
            blk_eve = Event(time_blk,4,self.peers[i])
            self.push(blk_eve)
        self.adj = {i:set() for i in range(n)}
        self.rho = [[0.0 for _ in range(n)] for _ in range(n)]
        self.visited = [False for _ in range(n)]
        self.create_graph()

    # dfs
    def dfs(self, pid):
        '''depth first search traversal'''
        self.visited[pid] = True
        for i in self.adj[pid]:
            if not self.visited[i]:
                self.dfs(i)

    # create graph
    def create_graph(self):
        '''create graph'''
        n = self.n
        self.adj = {i:set() for i in range(n)}
        conn = False
        while not conn:
            self.visited = [False for _ in range(n)]
            self.dfs(0)
            conn = True
            for i in range(n):
                if not self.visited[i]:
                    conn = False
                    break
            if conn:
                break
            done = False
            while not done:
                done = True
                self.adj = {i:set() for i in range(n)}
                for i in range(n):
                    req_ngbrs = np.random.choice(range(self.m,self.M + 1)) - len(self.adj[i])
                    if req_ngbrs < 0:
                        done = False
                        break
                    new_ngbrs = np.random.choice(n,req_ngbrs)
                    for j in new_ngbrs:
                        if j == i or len(self.adj[j]) > self.M:
                            done = False
                            break
                        self.adj[i].add(j)
                        self.adj[j].add(i)
                    for j in range(n):
                        if len(self.adj[j]) < self.m:
                            done = False
                            break
                if not done:
                    break
        for i in range(n):
            for j in self.adj[i]:
                self.rho[i][j] = np.random.uniform(10.0,500.0) * 0.001

    # print graph
    def print_graph(self):
        '''print graph'''
        n = self.n
        for i in range(n):
            print(f'{i}: ', end = '')
            for j in self.adj[i]:
                print(f'{j}, ', end = '')
            print()

    # runner
    def run(self):
        '''runner function'''
        while not len(self.prioq) == 0:
            eve = self.prioq.pop()
            if eve.timestamp > self.T and (eve.type == 1 or eve.type == 4):
                continue
            eve.run(self)

    # push events
    def push(self, eve):
        '''push events'''
        self.prioq.append(eve)
        self.prioq = sorted(self.prioq, key = lambda x: x.timestamp)

    # print entire tree
    def print_tree(self, fptr:TextIOWrapper):
        '''print tree'''
        for bid in range(Blk.curr_blk_id):
            blk = Blk.blk_i2p[bid]
            if blk is None:
                continue
            for child in blk.children:
                if child is None:
                    continue
                fptr.write(f'id_{bid}_miner_')
                if blk.miner is None:
                    fptr.write('G_')
                else:
                    fptr.write(f'{blk.miner.pid}_')
                fptr.write(f'inv_{blk.invalid} ')
                fptr.write(f'id_{child.blk_id}_miner_{child.miner.pid}_invalid_{child.invalid}\n')

    # print lengths
    def print_lengths(self, fptr:TextIOWrapper):
        '''print branch lengths'''
        for bid in range(Blk.curr_blk_id):
            blk = Blk.blk_i2p[bid]
            if blk is None:
                continue
            if len(blk.children) == 0:
                fptr.write(f'{blk.height}\n')
